### ISSUE TITLE
HiggsBounds interface

### ISSUE THREAD

**Issue by wkotlarski** [FlexibleSUSY/FlexibleSUSY#253]
The interface could be meta independent and look something like that
```cpp
template<typename Model_decays>    
void write_HB(Model_decays const& decays) {    
    
   const auto decay_table = decays.get_decay_table();   
   ....
```
The decay table can be search for specific decay by PDG identifiers. One missing thing are SM decays. Could it be possible to extend build in SM to include decays? I need it for the normalization of effective couplings. I would either pass it side by side to `write_HB` (or however we will call it) or even better, create it directly inside of `write_HB` function.

**Comment by Expander** [comment]
This interface looks simple, so that's nice!

From my point, we could of course extend the code in `model_specific/SM/` to add the SM decays. Note, that you cannot directly use code that you've generated with SARAH's SM for the inclusion in `model_specific/SM/`, because of various sign differences in the Lagrangian parameters. The code in `model_specific/SM/` has been generated with an old SARAH version, where the signs of some parameters (Yukawa couplings etc.) were different.

**Comment by wkotlarski** [comment]
This will be a problem. The decays don't care about models's convention. So that's fine. But decays need `CXXVertices`. Currently, the `standard_model` doesn't contain anything from Jobst's `cxx_qft` stuff.

**Comment by Expander** [comment]
Yes, so we'd have to (re-)generate large pieces of the `standard_model` code, including Jobst's pieces for the decays. I think this is in principle feasible, but great care has to be taken in order to not introduce bugs.

**Comment by pathron** [comment]
So when I was running FS this with LILITH I think I was just calling the autogenerated SM get the effective couplings for SM  for the ratios, so that is the option, ie  let it be done on the fly automatically.  I agree for efficiency and also cleaner setup it would be nicer to do it with model_specific/SM though.

**Comment by wkotlarski** [comment]
To summarise. This is a skeleton of potential function
```cpp
template<typename Model_decays>    
void write_HB(Model_decays const& decays) {    
    
   // identify SM-like Higgs in a BSM model
   // create SM with a Higgs of mass equal to the one in a BSM model
   // calculate decays in both models
   
   const auto decay_table_bsm = decays_bsm.get_decay_table();
   const auto decay_table_sm = decays_sm.get_decay_table();    

   // loop over final states printing ratios of widths
```

**Comment by wkotlarski** [comment]
Is it possible to cast `standard_model::Standard_model` to `SM_mass_eigenstates`?

**Comment by wkotlarski** [comment]
The `decay` constructor expects as argument `SM_mass_eigenstates`. If I'd like to hardcode decays in `model_specific/SM`, would it be possible to just replace `SM_mass_eigenstates` with `standard_model::Standard_model`?

**Comment by Expander** [comment]
I'm not sure. The interface seems to be the same, so you may simply try. :) You may want to add a test to compare that it gives the same results as FS's `SM` model.

However, I remember one physical difference: I think the convention for the choice of the `mu2` and the `Lambda` parameters is different than for FS's `SM` model. That is something to keep in mind when filling the input parameters.

**Comment by wkotlarski** [comment]
It's not a matter of simply trying. I'll have to edit the code by hand so I wanted to first consult it. I'll see if it works.

**Comment by wkotlarski** [comment]
@Expander, @pathron We will probably now want an interface to HiggsBounds/HiggsSignals replacement - [HiggsTools](https://gitlab.com/higgsbounds/higgstools). It says it requires a c++17 compliant compiler. Do you know if this sort of things also concerns clients (codes that call a given library). What I mean is: can we compile FS with c++14 compiler a still let it call a c++17 requiring library?

**Comment by Expander** [comment]
Excellent question!
I think this depends on whether the HiggsTools *header* (that we include) is C++14-compliant. If it is, then we can use it. (I'm assuming that linking is no problem here.)

**Comment by wkotlarski** [comment]
Thanks @Expander. That actually helped. So the answer is no. We will not be able to use it if we don't move to C++17. The problem is that one of the header uses `std::optional`
```sh
In file included from ../include/Higgs/predictions/ReferenceModels.hpp:16:
../include/Higgs/predictions/Particle.hpp:141:18: error: no template named 'optional' in namespace 'std'
    virtual std::optional<double> coupling(Coupling c) const noexcept;
```
Should we start thinking about transitioning to C++17 or will @pathron object?

**Comment by Expander** [comment]
Yes, let's discuss about switching to C++17! The supported compiler versions can be found here:

https://en.cppreference.com/w/cpp/compiler_support#cpp17

According to the table we need g++ >= 7 and clang++ >= 6.

* Debian stable: g++ 10.2, clang++ 11.0
* Ubuntu LTS (focal): g++ 9.3, clang++ 10
* Scientific linux 7: g++ 4.8.5, clang 
* Fedora: ?
* FreeBSD: g++ 10.4, clang++

So, it seems we'd be on the save side when switching to C++17 compatibility.

**Comment by pathron** [comment]
Scientific Linux is way behind though.  Also that’s a list of the support from the latest OS’s right? But most people don’t have the latest, we want to think about 3 or so years back so we support  what people use .  I have g++-7.5 on Ubuntu 18.04 though, so from there it’s ok.  Sorry to suggest something complicated but as an alternative is there anyway to enable the interface at configure only if the user  has g++-7 or higher?

**Comment by Expander** [comment]
Yes, this may be an option.

Alternatively, we could also create a C++14-compliant wrapper, which does not need `std::optional` etc. (Maybe this would be a good idea anyway, because it would allow us to create our own stable interface.)

**Comment by wkotlarski** [comment]
Closed by #446.
